src/datasets/table.py

```diff
@@ -558,53 +562,89 @@ def __setstate__(self, state):
-            tables_to_concat_vertically.append(combined_table)
-        return pa.concat_tables(tables_to_concat_vertically)
+            pa_table_horizontally_concatenated = cls._concat_blocks(tables, axis=1)
+            pa_tables_to_concat_vertically.append(pa_table_horizontally_concatenated)
+        return cls._concat_blocks(pa_tables_to_concat_vertically, axis=0)
+
+    @classmethod
+    def _merge_blocks(cls, blocks: TableBlockContainer, axis: Optional[int] = None) -> TableBlockContainer:
+        if axis is not None:
+            merged_blocks = []
+            for is_in_memory, block_group in groupby(blocks, key=lambda x: isinstance(x, InMemoryTable)):
+                if is_in_memory:
+                    block_group = [InMemoryTable(cls._concat_blocks(list(block_group), axis=axis))]
+                merged_blocks += list(block_group)
+        else:  # both
+            merged_blocks = [cls._merge_blocks(row_block, axis=1) for row_block in blocks]
+            if all(len(row_block) == 1 for row_block in merged_blocks):
+                merged_blocks = cls._merge_blocks(
+                    [block for row_block in merged_blocks for block in row_block], axis=0
+                )
+        return merged_blocks
+
+    @classmethod
+    def _consolidate_blocks(cls, blocks: TableBlockContainer) -> TableBlockContainer:
+        if isinstance(blocks, TableBlock):
+            return blocks
+        elif isinstance(blocks[0], TableBlock):
+            return cls._merge_blocks(blocks, axis=0)
+        else:
+            return cls._merge_blocks(blocks)
 
     @classmethod
     def from_blocks(cls, blocks: TableBlockContainer) -> "ConcatenationTable":
+        blocks = cls._consolidate_blocks(blocks)
         if isinstance(blocks, TableBlock):
             table = blocks
             return cls(table.table, [[table]])
         elif isinstance(blocks[0], TableBlock):
-            table = cls._concat_blocks_vertically(blocks)
+            table = cls._concat_blocks(blocks, axis=0)
             blocks = [[t] for t in blocks]
             return cls(table, blocks)
         else:
             table = cls._concat_blocks_horizontally_and_vertically(blocks)
             return cls(table, blocks)
 
     @classmethod
-    def from_tables(cls, tables: List[Union[pa.Table, Table]]) -> "ConcatenationTable":
+    def from_tables(cls, tables: List[Union[pa.Table, Table]], axis: int = 0) -> "ConcatenationTable":
         blocks = []
         for table in tables:
             if isinstance(table, pa.Table):
-                blocks.append([InMemoryTable(table)])
+                blocks.append([InMemoryTable(table)] if axis == 0 else InMemoryTable(table))
             elif isinstance(table, ConcatenationTable):
-                blocks.extend(copy.deepcopy(table.blocks))
+                blocks.extend(copy.deepcopy(table.blocks) if axis == 0 else copy.deepcopy(table.blocks)[0])  # TODO
             else:
```

Comment:
@lhoestq here we have a potential issue if we want to concatenate horizontally a block `A` with a ConcatenationTable with multiple block rows `[[B], [C]]`. The resulting blocks should be something like `[[A_1, B], [A_2, C]]`.

For the moment I only concatenate horizontally to the first block row.

Reply:
Blocks of a single row should have the same length or the column concatenation fails.
So ideally we should end up with something like `[[A_1, B], [A_2, C]]` as you mentioned.
Feel free to construct `A_1` and `A_2` by calling `A.slice(offset, length)`.

Reply:
Done.